### PR TITLE
Reinstating old behaviour of += with lists

### LINKED ISSUE #1152: Shapes.scl broken by new method of adding lists and sets via +=
Now `list_variable += set` will make the set an element of the list, as opposed to previously when it would add all the elements. While this is good in most cases, it screws up. I have a solution for that though, and hear it out.

I propose to add streams, or sequentially generated objects. The way they will work is that you call `next(stream)` and it will evaluate the next item, and you declare them via `stream = stream(_(prevstate)->{'ret'->prevstate:'acc'+1, 'acc' -> prevstate:'acc'+1})` 
So basically, the `prevstate` is a map of the previous state, and you can set whatever state you want, holding whatever variables yo want as keys, and then there is a `'ret'` element of the map (can be renamed `return`, whatevs) which means whenever you do like:
`x = next(stream)`, x will be assigned to that `ret` parameter. This would fix it as then I could sequentially generate the lists and simultaneously add them .

Now after further thought, the above problem can be fixed using `x = x + set` isntead of `+=`, but I'm still gonna leave this stream object stuff here, cos I think it's still interesting. I will submit a pr to fix the shapes.scl problem, then make another issue to discuss the stream stuff.

### PR DESCRIPTION
**Pull request by Ghoulboy78**
Properly fixes #1152, as opposed to the hack-fix in #1154 which just works around it.
In ye olden days, when the += operator simply called `Value.add()`, this is the behaviour that would occur, i.e.: doing `l=[]; l+=['3','4']`, would result in `l==['3','4']`, whereas now it results in `l==[['3','4']]` . This broke `shapes.scl` accidentally. I have also fixed this behaviour for maps, cos it's also broken there.

### CODE REVIEW

**Comment by ch-yx** [comment]
Oh no. 🤦‍
Do you think the semantics of += are not ambiguous enough?

**Comment by Firigion** [comment]
I think that if you want to use `+=` on lists to mean `lhs = lhs + rhs`, you need to also add an easy way to append into a list. `+=` was added as append because  `put(my_list, null, element)` is way to verbose for sucha a common action. I think `+=` functioning as append is a fantastic addition and old code that used the fact that adding lists of the same length wolud add them element-wise just need to switch their `l += [1,2]` to `l = l+[1,2]`, which is not that bad.

Also, `l=[]; l+=['3','4']` returning `l==['3','4']` also doesn't make much sense. If addition between lists is meant to add element-wise, then this should error out. If it's meant to _extend_ a list (append all the contents of the right hand side list into the end of the left hand side one), this is gonna silently fail when the two lists are the same length. I'd strongly advice against removing element-wise operations of lists, since we dont' ahve a better data type for that.

Basically, this bolis down to what is more commonly used: extend or append. One should keep the `+=` action and the other one stick to `put(list, null, value, mode)`. My vote is for append to keep it.

**Comment by Ghoulboy78** [comment]
Ok, well what I did was make it so if you do `l += a_list_or_map`, the elements of `a_list_or_map` (for a map it's the key-value pairs in a list, like you used to declare them in the olden days) get added one by one to the end of list `l`. For anything else though, doing `l+=v` extends the list and sticks `v` to the end of `l`. But you can still do element-wise addition with `l = l+a_list` (not with a map I don't think). Basically I didn't touch `+` operator, just `+=` for lists, which is called via the `append()` method in `AbstractListValue` class.

**Comment by Ghoulboy78** [comment]
But you see Firi, I'm reinstating old behaviour which was accidentally removed earlier on, and no one noticed that it broke stuff.

**Comment by Firigion** [comment]
I know, but I don't like it, and I assume gnemon changed it on purpose :P

Still, if you are making a PR reisntating the old behaviour, you can use the opportunity to clarify the thing to new user with a good documentation, because the behaviour is anything but trivial.

**Comment by Ghoulboy78** [comment]
I made a discord thread to discuss it

**Comment by gnembon** [comment]
Yeah - current behaviour is less confusing then proposed. 'put' has clear usecases for 'insert' and 'extend'. Plus that would break a few examples that are currently in the docs and work because += assume the value to add is always a scalar.
